server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala

```diff
@@ -209,16 +211,43 @@ class SparkYarnApp private[utils] (
       listener.foreach(_.appIdKnown(appId.toString))
 
       val pollInterval = SparkYarnApp.getYarnPollInterval(livyConf)
+      var appInfo = AppInfo()
       while (isRunning) {
-        Clock.sleep(pollInterval.toMillis)
-
-        // Refresh application state
-        val appReport = yarnClient.getApplicationReport(appId)
-        yarnDiagnostics = getYarnDiagnostics(appReport)
-        changeState(mapYarnState(
-          appReport.getApplicationId,
-          appReport.getYarnApplicationState,
-          appReport.getFinalApplicationStatus))
+        try {
+          Clock.sleep(pollInterval.toMillis)
+
+          // Refresh application state
+          val appReport = yarnClient.getApplicationReport(appId)
+          yarnDiagnostics = getYarnDiagnostics(appReport)
+          changeState(mapYarnState(
+            appReport.getApplicationId,
+            appReport.getYarnApplicationState,
+            appReport.getFinalApplicationStatus))
+
+          val latestAppInfo = {
+            val attempt =
+              yarnClient.getApplicationAttemptReport(appReport.getCurrentApplicationAttemptId)
+            val driverLogUrl =
+              Try(yarnClient.getContainerReport(attempt.getAMContainerId).getLogUrl)
```

Comment:
Does it mean driver should be run on AM, here the driver log is actually AM log? So in that case, how can we handle yarn client mode, or other cluster manager like Standalone or Mesos?